File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        start, end = newInterval
        ans = []
        s_index = 0
        for i in range(len(intervals)):
            if intervals[i][0] > start:
                s_index = i 
                break
            else:
                ans.append(intervals[i])
        #print("1: ", ans)
        if ans:
            if start <= ans[-1][1]:
                if ans[-1][1] < end:
                    ans[-1][1] = end
            else:
                ans.append(newInterval)
        else:
            ans = [newInterval]

        #print("2: ", ans)
        for i in range(s_index, len(intervals)):
            if intervals[i][0] <= ans[-1][1]:
                if ans[-1][1] <= intervals[i][1]:
                    ans[-1][1] = intervals[i][1]
            else:
                ans += intervals[i:]
                break
        return ans

File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        intervals.push_back(newInterval);
        sort(intervals.begin(), intervals.end());
        stack<vector<int>> s;
        vector<vector<int>> ans;
        for (int i = 0; i < intervals.size(); i++) {
            if (s.empty()) {
                s.push(intervals[i]);
            } else {
                vector<int> curr = s.top();
                if (intervals[i][0] > curr[1]) {
                    ans.push_back(curr);
                    s.pop();
                    s.push(intervals[i]);
                } else if (curr[1] < intervals[i][1]){
                    s.top()[1] = intervals[i][1];
                }
            } 
        }
        if (!s.empty()) {
            ans.push_back(s.top());
        }
        return ans;
    }
};